tools/cgcollector2/src/CallGraphCollectionAction.cpp

```diff
@@ -0,0 +1,182 @@
+/**
+ * File: CallGraphCollectionAction.cpp
+ * License: Part of the MetaCG project. Licensed under BSD 3 clause license. See LICENSE.txt file at
+ * https://github.com/tudasc/metacg/LICENSE.txt
+ */
+
+#include <fstream>
+#include <iostream>
+
+#include "MCGManager.h"
+#include "io/VersionTwoMCGWriter.h"
+#include "io/VersionFourMCGWriter.h"
+
+#include "CallGraphCollectionAction.h"
+#include "CallGraphNodeGenerator.h"
+#include "Plugin.h"
+#include "SharedDefs.h"
+#include "metadata/Internal/ASTNodeMetadata.h"
+#include "metadata/Internal/AllAliasMetadata.h"
+#include "metadata/Internal/FunctionSignatureMetadata.h"
+
+#include "clang/AST/ASTDumper.h"
+
+void CallGraphCollectorConsumer::HandleTranslationUnit(clang::ASTContext& Context) {
+  std::string dumpString;
+  llvm::raw_string_ostream llvmStringOstream(dumpString);
+  Context.getTranslationUnitDecl()->dump(llvmStringOstream);
+  SPDLOG_TRACE("\n{}\n", dumpString);
+
+  // Todo: maybe use multigraph managing and write only after all files have been processed
+  auto& mcgm = metacg::graph::MCGManager::get();
+  mcgm.resetManager();
+  mcgm.addToManagedGraphs("newGraph", std::make_unique<metacg::Callgraph>());
+  auto callgraph = mcgm.getCallgraph();
+  CallGraphNodeGenerator graphGenerator(callgraph, captureCtorsDtors, captureNewDeleteCalls, captureImplicits, inferCtorsDtors,
+                                        standalone, level);
+
+  graphGenerator.TraverseDecl(Context.getTranslationUnitDecl());
+
+  // Augment the graph with alias information if this the only TU we will get
+  // if other TUs are present, overestimation needs to be done during merge
+  if (level == AliasAnalysisLevel::All) {
+    addOverestimationEdges(callgraph);
+  }
+
+  SPDLOG_INFO("Sucessfully Created Callgraph");
+  SPDLOG_INFO("Running Metadata Collectors");
+  for (auto& c : mcs) {
+    SPDLOG_DEBUG("Running: {}", c->getPluginName());
+    for (const auto& node : callgraph->getNodes()) {
+      if (!node->has<ASTNodeMetadata>()) {
+        // This is because we currently do not attach ast node metadata to function pointers
+        continue;
+      }
+      auto nodeDecl = node->get<ASTNodeMetadata>()->getFunctionDecl();
+      auto md = c->computeForDecl(nodeDecl);
+      if (md) {
+        node->addMetaData(std::move(md));
+      }
+    }
+    c->computeForGraph(callgraph);
+  }
+
+  std::unique_ptr<metacg::io::MCGWriter> mcgWriter;
+
+  switch (mcgVersion){
+    case 1: SPDLOG_INFO("This tool can not generate output files in the V1 format, using V2 instead");
+      __attribute__ ((fallthrough));
+    case 2:
+      mcgWriter = std::make_unique<metacg::io::VersionTwoMCGWriter>();
+      break;
+    case 3: SPDLOG_INFO("V3 format was removed and is currently not supported, using V4 instead");
+      __attribute__ ((fallthrough));
+
+    case 4: mcgWriter = std::make_unique<metacg::io::VersionFourMCGWriter>();
+      break;
+    default:
+      assert(false && "No output format specified");
```

Comment:
Is this behavior intended here, i.e., fail in debug builds but output something in release builds?

Reply:
A hacky way would be to move it as a `define` (or whatever) into a header that gets included in both places.